Release notes 1.5.0 (#4864)
STEllAR-GROUP/hpx: docs/sphinx/releases/whats_new_1_5_0.rst

.. _hpx_1_5_0:

===========================
|hpx| V1.5.0 (?)
===========================

General changes
===============

> * We have added the C++20 features ``hpx::jthread`` and ``hpx::stop_token``. We

Reviewer: We should add a couple of higher-level sentences describing the main focus of this release, i.e. C++20 conformance, modularization, performance improvements, API unification, etc.

  have also added ``hpx::stable_sort`` based on Francisco Tapia's
  implementation.
* We have adapted existing sychronization primitives to C++20: ``barrier``,
  ``latch``, ``counting_semaphore``.

> * We have started using customization point objects to make algorithm extension

Reviewer: Please mention full C++20 conformance for the adapted algorithms.

  easier for the user. ``reduce``, ``copy``, ``all_of``/``any_of``/``none_of``,
  ``fill``, ``count``, ``destroy``, ``move`` are using those CPOs.  We also have
  adapted their corresponding ``hpx::ranges`` versions in this release,
  deprecating the ``hpx::parallel`` ones.
* In the perspective of continuing our work on making the hpx algorithms support
  the ranges, we have adapted ``hpx::traits::is_range`` and others to use
  ``hpx::traits::is_sentinel_for`` concept. We have adapted ``hpx::copy``
  and ``hpx::stable_sort`` algorithms.

> * We have adapted ``co_await`` to C++20, it now supports ``shared_future`` and

Reviewer: ```suggestion
* We have adapted support for ``co_await`` to C++20, it now supports ``shared_future`` and
```

  have added allocator support for futures returned by ``co_return``. It is no
  longer in the ``experimental`` namespace.
* We added serialization support for ``std::variant`` and ``std::tuple``.
* We continued with the modularization, it is now easier to enable/disable the
  distributed runtime option. An important amount of headers have been
  deprecated. We adapted the namespaces and headers we could to be closer to the
  standard ones (:ref:`api <_public_api>`).

> * We have set the HPX C++ standard automatically to C++14 and if it needs to be

Reviewer: This is a breaking change as HPX now requires C++14 as the minimal compilation mode.

Author: Yes I initially separated the change to the mimimal supported standard in the breaking change section and the new cmake option related to it in the general changes section but it makes more sense to have it as the same place, thanks :)

* We have added a ``hpx::init_params`` struct to pass parameters for HPX
  initialization e.g. the resource partitioner callback to initialize thread
  pools (:ref:`using the resource partitioner from hpx 1.5.0 onwards
  <_using_rp_from_1_5_0>`).
* A polymorphic executor has been implemented to reduce compile times as a
  function accepting executor can potentially be instanciated only once instead
  of multiple times with different executors. It accepts the function signature
  as a template argument. It needs to be constructed from any other executor.
  Please note, that the function signatures that can be scheduled using
  ``then_execute``, ``bulk_sync_execute``, ``bulk_async_execute`` and
  ``bulk_then_execute`` are slightly different.
* The underlying executor of ``block_executor`` has been updated to a newer one.
* We have added a parameter to ``auto_chunk_size`` to control the amount of
  iterations to measure.
* We have added performance counters type information (aggregating,
  monotonically increasing, average count, average timer, etc.).
* The tasks are now resumed on the worker thread they were suspended to avoid

>   cache misses from moving from one thread to the other.

Reviewer: Is this a general rule? That would essentially disable work-stealing. Or did you mean to say something different?

Reviewer 2: Yeah, maybe this wording is slightly too strong. Tasks are re-*scheduled* on the same worker thread where they were suspended. It doesn't stop the task from being stolen though if the worker thread doesn't get to executing it before another worker thread gets that far.

Author: Is this ok?
```
+* The tasks are now re-scheduled on the same worker thread they were suspended
+  on to avoid cache misses from moving from one thread to the other. This
+  behavior doesn't prevent the task from being stolen."
```

Reviewer 2: Sounds good to me. Thanks!

* We have added a new configuration option ``hpx.exception_verbosity`` to allow
  to control the level of verbosity of the exceptions (3 levels available).
* ``broadcast_to``, ``broadcast_from``, ``scatter_to`` and ``scatter_from`` have
  been added to the collectives, modernization of ``gather_here`` and
  ``gather_there`` with futures taken by rvalue references, rename the existing
  ``all_to_all`` to ``all_gather`` and adds a new implementation of

>   ``all_to_all``.

Reviewer: All of the collectives do not need supporting macros anymore (e.g. specifying the data types used for a collective operation using `HPX_REGISTER_ALLGATHER` and similar are no needed anymore).

* A new API function to get the number of cores which are idle

>   (``hpx::get_idle_core_count``)

Reviewer: There is also `get_idle_core_mask` returning a bitmask representing the currently idle cores.

* We have added an experimental option to only enable the local runtime, you can
  enable the distributed runtime with ``HPX_WITH_DISTRIBUTED_RUNTIME=ON`` (or
  ``OFF`` to disable it and have only the local one). You can also enable the
  local runtime by using the ``--hpx:local`` runtime option.
* We fixed task annotations for actions.

> * ``hpx::promise`` is now deprecated and transformed into ``hpx::lcos::promise``.

Reviewer: Huh? I think this needs to be rephrased.

Author: indeed ^^'

* ``result_of`` and ``is_callable`` are now deprecated and replaced by

>   ``invoke_result``.

Reviewer: ... and `is_invocable` (i.e. conforming to C++20)

* We have added a ``prepare_checkpoint`` API function which calculates the
  amount of necessary buffer space for a particular set of argumnets
  checkpointed.

> * We have changed the exposed targets to the user, it now includes ``HPX::hpx``,

Reviewer: ```suggestion
* We have changed the exposed cmake targets to the user, it now includes ``HPX::hpx``,
```

 ``HPX::wrap_main`` (``int main`` as the first HPX::thread of the application,
 see :ref:`starting HPX runtime <_starting_hpx>`),
 ``HPX::plugin``, ``HPX::component``.
* A new example is demonstrating how to create and use a wrapping executor
  (``quickstart/executor_with_thread_hooks.cpp``)
* A new example is demonstrating how to disable thread stealing during the
  execution of parallel algorithms
  (``quickstart/disable_thread_stealing_executor.cpp``)
* Many bug fixes.

Breaking changes
================

> * HPX now requires C++14 capable compiler.

Reviewer: also:

- runtime initialization using `resource_partitioner`
- `all_to_all` is now called `all_gather`, new `all_to_all` is not compatible with old

  ``shared_future`` and have added allocator support for futures returned by
  ``co_return``. It is no longer in the ``experimental`` namespace.
* We added serialization support for ``std::variant`` and ``std::tuple``.
* ``result_of`` and ``is_callable`` are now deprecated and replaced by
  ``invoke_result`` and ``is_invocable`` to conform to C++20.
* We continued with the modularization, making it easier for us to add the new
  experimental ``HPX_WITH_DISTRIBUTED_RUNTIME`` |cmake| option (see below) . An
  important amount of headers have been deprecated. We adapted the namespaces
  and headers we could to be closer to the standard ones (:ref:`api
  <_public_api>`).
* It is now possible to have a basic CUDA support including a helper function to
  get a future from a CUDA stream and target handling. They are available under
  the ``hpx::cuda::experimental`` namespace and they can be enabled with the
  ``HPX_WITH_ASYNC_CUDA=ON`` cmake option.
* We added a new ``hpx::mpi::experimental`` namespace for getting futures from
  an asynchronous MPI call and a new minimal MPI executor (add the name).
* A polymorphic executor has been implemented to reduce compile times as a

>   function accepting executor can potentially be instanciated only once instead

Reviewer 2: ```suggestion
  function accepting executor can potentially be instantiated only once instead
```

  can also enable the local runtime by using the ``--hpx:local`` runtime option.
* We fixed task annotations for actions.
* The alias ``hpx::promise`` to ``hpx::lcos::promise`` is now deprecated. You
  can use ``hpx::lcos::promise`` directly instead. ``hpx::promise`` will refer
  to the local-only promise in the future.
* We have added a ``prepare_checkpoint`` API function which calculates the
  amount of necessary buffer space for a particular set of argumnets
  checkpointed.
* We have changed the |cmake| targets exposed to the user, it now includes
  ``HPX::hpx``, ``HPX::wrap_main`` (``int main`` as the first |hpx| thread of the
  application, see :ref:`starting HPX runtime <_starting_hpx>`),
  ``HPX::plugin``, ``HPX::component``.
* A new example is demonstrating how to create and use a wrapping executor
  (``quickstart/executor_with_thread_hooks.cpp``)
* A new example is demonstrating how to disable thread stealing during the
  execution of parallel algorithms
  (``quickstart/disable_thread_stealing_executor.cpp``)
* Many bug fixes.

Breaking changes
================

* |hpx| now requires C++14 capable compiler. We have set the |hpx| C++ standard

> automatically to C++14 and if it needs to be set explicitly, it should be

Reviewer 2: ```suggestion
  automatically to C++14 and if it needs to be set explicitly, it should be
```

conformant. As part of the modularization we have begun specifying the public
API of HPX in terms of headers and functionality, and aligning it more closely
to the C++ standard. All non-distributed modules are now in place, along with an
experimental option to completely disable distributed features in HPX. We have
also added an experimental asynchronous MPI executor. Lastly this release
contains official |cmake| targets, performance improvements, and many bug fixes.

* We have added the C++20 features ``hpx::jthread`` and ``hpx::stop_token``. We
  have also added ``hpx::stable_sort`` based on Francisco Tapia's
  implementation.
* We have adapted existing sychronization primitives to be fully conformant
  C++20: ``barrier``, ``latch``, ``counting_semaphore``.
* We have started using customization point objects to make the corresponding
  algorithms fully conformant to C++20 as well as to make algorithm extension
  easier for the user. ``reduce``, ``copy``, ``all_of``/``any_of``/``none_of``,
  ``fill``, ``count``, ``destroy``, ``move`` are using those CPOs.  We also have
  adapted their corresponding ``hpx::ranges`` versions in this release,

>   deprecating the ``hpx::parallel`` ones.

Reviewer 2: The `hpx::parallel` ones will be deprecated whenever all the algorithms have been adapted (hopefully next release), so you can remove the last part.

> * In the perspective of continuing our work on making the hpx algorithms support

Reviewer 2: ```suggestion
* In the perspective of continuing our work on making the |hpx| algorithms support
```
and potentially elsewhere as well.

  the ranges, we have adapted ``hpx::traits::is_range`` and others to use
  ``hpx::traits::is_sentinel_for`` concept. We have adapted ``hpx::copy`` and

>   ``hpx::stable_sort`` algorithms.

Reviewer 2: Are these the only algorithms that have been adapted? It may be easier to leave out the list here if it's just the same list as above.

* We have adapted support for ``co_await`` to C++20, it now supports
  ``shared_future`` and have added allocator support for futures returned by
  ``co_return``. It is no longer in the ``experimental`` namespace.
* We added serialization support for ``std::variant`` and ``std::tuple``.
* ``result_of`` and ``is_callable`` are now deprecated and replaced by
  ``invoke_result`` and ``is_invocable`` to conform to C++20.

> * We continued with the modularization, it is now easier to enable/disable the

Reviewer 2: ```suggestion
* We continued with the modularization, making it easier for us to add the new experimental ``HPX_WITH_DISTRIBUTED_RUNTIME`` |cmake| option (see below).
```

  distributed runtime option. An important amount of headers have been
  deprecated. We adapted the namespaces and headers we could to be closer to the
  standard ones (:ref:`api <_public_api>`).
* It is now possible to have a basic CUDA support including a helper function to
  get a future from a CUDA stream and target handling. They are available under
  the ``hpx::cuda::experimental`` namespace and they can be enabled with the
  ``HPX_WITH_ASYNC_CUDA=ON`` cmake option.
* We added a new ``hpx::mpi::experimental`` namespace for getting futures from
  an asynchronous MPI call and a new minimal MPI executor (add the name).
* A polymorphic executor has been implemented to reduce compile times as a
  function accepting executor can potentially be instanciated only once instead
  of multiple times with different executors. It accepts the function signature
  as a template argument. It needs to be constructed from any other executor.
  Please note, that the function signatures that can be scheduled using
  ``then_execute``, ``bulk_sync_execute``, ``bulk_async_execute`` and

>   ``bulk_then_execute`` are slightly different.

Reviewer 2: Slightly different how?

Author: Should I detail the signatures here or refer to the corresponding [comment](https://github.com/STEllAR-GROUP/hpx/pull/4514#issuecomment-613139299) on Github?

Reviewer: A link to the github ticket is probably fine.

  monotonically increasing, average count, average timer, etc.).
* The tasks are now re-scheduled on the same worker thread they were suspended
  on to avoid cache misses from moving from one thread to the other. This
  behavior doesn't prevent the task from being stolen.
* We have added a new configuration option ``hpx.exception_verbosity`` to allow
  to control the level of verbosity of the exceptions (3 levels available).
* ``broadcast_to``, ``broadcast_from``, ``scatter_to`` and ``scatter_from`` have
  been added to the collectives, modernization of ``gather_here`` and
  ``gather_there`` with futures taken by rvalue references. See the breaking
  change on ``all_to_all`` in the next section. All the collectives do not need
  supporting macros anymore (e.g. specifying the data types used for a
  collective operation using ``HPX_REGISTER_ALLGATHER`` and similar are not
  needed anymore).
* A new API function to get the number of cores which are idle
  (``hpx::get_idle_core_count``) and another one returning a bitmask
  representing the currently idle cores (``get_idle_core_mask``).
* We have added an experimental option to only enable the local runtime, you can

>   enable the distributed runtime with ``HPX_WITH_DISTRIBUTED_RUNTIME=ON`` (or

Reviewer 2: I'd maybe turn this around and say "you can disable the distributed runtime at compile time with `HPX_WITH_DISTRIBUTED_RUNTIME=OFF` (defaults to `ON`)"?

  can use ``hpx::lcos::promise`` directly instead. ``hpx::promise`` will refer
  to the local-only promise in the future.
* We have added a ``prepare_checkpoint`` API function which calculates the
  amount of necessary buffer space for a particular set of argumnets
  checkpointed.
* We have changed the |cmake| targets exposed to the user, it now includes
  ``HPX::hpx``, ``HPX::wrap_main`` (``int main`` as the first |hpx| thread of the
  application, see :ref:`starting HPX runtime <_starting_hpx>`),
  ``HPX::plugin``, ``HPX::component``.
* A new example is demonstrating how to create and use a wrapping executor
  (``quickstart/executor_with_thread_hooks.cpp``)
* A new example is demonstrating how to disable thread stealing during the
  execution of parallel algorithms
  (``quickstart/disable_thread_stealing_executor.cpp``)
* Many bug fixes.

> 

Reviewer: There is one more thing added: all executor parameter hooks can now be exposed through the executor itself. This will allow to deprecate the `.with()` functionality on execution policies in the future. This is also a first step towards simplifying our executor APIs in preparation for the upcoming C++23 executors (senders/receivers).

Author: I didn't bother to rephrase as your name is on the copyright and my english is pretty bad, I hope you don't mind ;)

Breaking changes
================

> * |hpx| now requires C++14 capable compiler. We have set the |hpx| C++ standard

Reviewer: ```suggestion
* |hpx| now requires a C++14 capable compiler. We have set the |hpx| C++ standard
```
